common/compositor/factory.cpp

```diff
@@ -17,46 +17,106 @@
 #include "factory.h"
 #include "platformdefines.h"
 
-#ifdef USE_GL
 #include "glrenderer.h"
 #include "glsurface.h"
 #include "nativeglresource.h"
-#elif USE_VK
 #include "nativevkresource.h"
 #include "vkrenderer.h"
 #include "vksurface.h"
-#endif
 
 namespace hwcomposer {
 
-NativeSurface* CreateBackBuffer(uint32_t width, uint32_t height) {
-#ifdef USE_GL
+static NativeSurface* create_gl_surface(uint32_t width, uint32_t height) {
   return new GLSurface(width, height);
-#elif USE_VK
+}
+
+static Renderer* create_gl_renderer() {
+  return new GLRenderer();
+}
+
+static NativeGpuResource* create_gl_resource() {
+  return new NativeGLResource();
+}
+
+static NativeSurface* create_vk_surface(uint32_t width, uint32_t height) {
   return new VKSurface(width, height);
+}
+
+static Renderer* create_vk_renderer() {
+  return new VKRenderer();
+}
+
+static NativeGpuResource* create_vk_resource() {
+  return new NativeVKResource();
+}
+
+struct backend {
+  const char* id;
+  bool (*is_supported)();
+  NativeSurface* (*create_surface)(uint32_t width, uint32_t height);
+  Renderer* (*create_renderer)();
+  NativeGpuResource* (*create_resource)();
+};
+
+static struct backend backends[] = {
+#ifdef USE_VK
+    {"VK", vk_is_supported, create_vk_surface, create_vk_renderer,
+     create_vk_resource},
+    {"GL", gl_is_supported, create_gl_surface, create_gl_renderer,
+     create_gl_resource},
```

Comment:
Compile failed on Android whether USE_VK is enabled or not. 

As backend is selected with different priority here, both vk and gl source codes are need to compile. 
BOARD_USES_VULKAN switch should be removed from Android.mk, and vk and gl comiple options need to merge together.